library/src/math/matrix.rs

```diff
@@ -242,35 +242,29 @@ fn layout_mat_body(ctx: &mut MathContext, rows: &[Vec<Content>]) -> SourceResult
         return Ok(Frame::new(Size::zero()));
     }
 
-    let mut widths = vec![Abs::zero(); ncols];
-    let mut ascents = vec![Abs::zero(); nrows];
-    let mut descents = vec![Abs::zero(); nrows];
+    let mut heights = vec![(Abs::zero(), Abs::zero()); nrows];
 
     ctx.style(ctx.style.for_denominator());
     let mut cols = vec![vec![]; ncols];
-    for ((row, ascent), descent) in rows.iter().zip(&mut ascents).zip(&mut descents) {
-        for ((cell, rcol), col) in row.iter().zip(&mut widths).zip(&mut cols) {
+    for (row, (ascent, descent)) in rows.iter().zip(&mut heights) {
+        for (cell, col) in row.iter().zip(&mut cols) {
             let cell = ctx.layout_row(cell)?;
-            rcol.set_max(cell.width());
             ascent.set_max(cell.ascent());
             descent.set_max(cell.descent());
             col.push(cell);
         }
     }
     ctx.unstyle();
 
-    let width = widths.iter().sum::<Abs>() + col_gap * (ncols - 1) as f64;
-    let height = ascents.iter().sum::<Abs>()
-        + descents.iter().sum::<Abs>()
-        + row_gap * (nrows - 1) as f64;
-    let size = Size::new(width, height);
-
-    let mut frame = Frame::new(size);
+    let mut frame = Frame::new(Size::new(
+        Abs::zero(),
+        heights.iter().map(|&(a, b)| a + b).sum::<Abs>() + row_gap * (nrows - 1) as f64,
+    ));
     let mut x = Abs::zero();
-    for (col, &rcol) in cols.into_iter().zip(&widths) {
-        let points = alignments(&col);
+    for col in cols {
+        let AlignmentResult { points, width: rcol } = alignments(&col);
```

Comment:
Computing the width by taking alignment into account is the fix. The other cleanups are because we don't need to compute width twice anymore.